lib/rules/lint-no-attr-in-components.js

```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+const Rule = require('./base');
+
+module.exports = class LinkNoAttrComponent extends Rule {
+  isComponentTemplate() {
+    let moduleName = this.templateEnvironmentData.moduleName;
+    let regex = RegExp('templates/components/|components/.*/template.hbs');
+    return regex.test(moduleName);
+  }
+  logAttrsInComponentTemplates(node, attributes, identifier) {
+    attributes.forEach((attribute) => {
+      let value = identifier ? attribute[identifier] : attribute;
+      if (value.parts && value.parts[0] === 'attrs') {
+        this.log({
+          message: 'Component templates should not contain `attrs`.',
+          line: value.loc && value.loc.start.line,
+          column: value.loc && value.loc.start.column,
+          source: this.sourceForNode(node)
+        });
+      }
+    });
+  }
+  processParamsAndHashPairs(node) {
+    let attributes = (node.hash || {}).pairs;
+    let params = node.params || [];
+
+    if (!this.isComponentTemplate()) {
+      return;
+    }
+    if (attributes && attributes.length || params.length) {
+      this.logAttrsInComponentTemplates(node, attributes, 'value');
+      this.logAttrsInComponentTemplates(node, params, '');
+    }
+  }
+  visitor() {
+    return {
+      SubExpression(node) {
```

Comment:
I think that you can likely get away with using `PathExpression` as the main hook here (instead of `SubExpression`, `BlockStatement`, `MustacheStatement` and not need to manually traverse `params` and `hash` of each)...

Reply:
The reason I did not use `PathExpression` was that I felt this rule should allow components to be named `attrs` or to be present in folders named `attrs` (although these cases aren't ideal)

Reply:
Hmm, how would using `PathExpression` prevent using components in folders named attrs?

Also, components already cannot be named exactly `attrs` due to the dash requirement...

Reply:
Sorry I first thought of components. But the actual problem would be with helpers named `attrs`

Reply:
We _could_ handle that while using path expression visitor (by only matching when `node.original.startsWith(‘attrs.’)`), but TBH I don’t think having a helper named attrs is common (or a good idea 😜).

I’d still much prefer to go the path expression route. It also would ensure compliance with angle bracket invocation (those would be `ElementNode`s)...

Reply:
Agreed. I guess I got caught up a bit with what the rule should and shouldn't support. I have made all the changes.
I have just one question though. Although this case may never ever come up, should we also guard against simply using `{{attrs}}` ?

Reply:
>  Although this case may never ever come up, should we also guard against simply using {{attrs}} ?

Hmm, good point. I agree, I don't think we want to allow that...

I think we basically want to log an error whenever `node.parts[0] === 'attrs` (for the `PathExpression` hook), what do you think? That should catch all of these:

* `{{attrs}}`
* `{{foo-bar blah=attrs}}`
* `{{attrs.derp}}`

Reply:
Yes it should work against all the cases and that includes the helpers. I guess we can go ahead with `node.parts[0]` check for now and see if the helper case really is a concern later on. Thoughts?

Reply:
Yep, totally agreed.